// ng2-charts/esm2015/lib/get-colors.js
import { baseColors } from './base-colors';
const ɵ0 = (context) => rgba(generateColor(context.datasetIndex), 0.4), ɵ1 = (context) => rgba(generateColor(context.datasetIndex), 1), ɵ2 = (context) => rgba(generateColor(context.datasetIndex), 1), ɵ3 = (context) => rgba(generateColor(context.datasetIndex), 0.8), ɵ4 = (context) => rgba(generateColor(context.datasetIndex), 0.6), ɵ5 = (context) => rgba(generateColor(context.datasetIndex), 1), ɵ6 = (context) => rgba(generateColor(context.datasetIndex), 0.8), ɵ7 = (context) => rgba(generateColor(context.datasetIndex), 1), ɵ8 = (context) => rgba(generateColor(context.dataIndex), 0.6), ɵ9 = (context) => rgba(generateColor(context.dataIndex), 1), ɵ10 = (context) => rgba(generateColor(context.dataIndex), 1), ɵ11 = (context) => rgba(generateColor(context.dataIndex), 0.6), ɵ12 = (context) => rgba(generateColor(context.dataIndex), 1), ɵ13 = (context) => rgba(generateColor(context.dataIndex), 0.8), ɵ14 = (context) => rgba(generateColor(context.dataIndex), 1);
export const builtInDefaults = {
    datasets: {
        line: {
            backgroundColor: ɵ0,
            borderColor: ɵ1,
            pointBackgroundColor: ɵ2,
            pointBorderColor: '#fff',
            pointHoverBackgroundColor: '#fff',
            pointHoverBorderColor: ɵ3
        },
        bar: {
            backgroundColor: ɵ4,
            borderColor: ɵ5,
            hoverBackgroundColor: ɵ6,
            hoverBorderColor: ɵ7
        },
        get radar() {
            return this.line;
        },
        doughnut: {
            backgroundColor: ɵ8,
            borderColor: '#fff',
            hoverBackgroundColor: ɵ9,
            hoverBorderColor: ɵ10
        },
        get pie() {
            return this.doughnut;
        },
        polarArea: {
            backgroundColor: ɵ11,
            borderColor: ɵ12,
            hoverBackgroundColor: ɵ13,
            hoverBorderColor: ɵ14
        },
        get bubble() {
            return this.doughnut;
        },
        get scatter() {
            return this.doughnut;
        },
        get area() {
            return this.polarArea;
        }
    }
};
function rgba(colour, alpha) {
    return 'rgba(' + colour.concat(alpha).join(',') + ')';
}
function getRandomInt(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}
function getRandomColor() {
    return [getRandomInt(0, 255), getRandomInt(0, 255), getRandomInt(0, 255)];
}
/**
 * Generate colors
 */
function generateColor(index = 0) {
    return baseColors[index] || getRandomColor();
}
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8, ɵ9, ɵ10, ɵ11, ɵ12, ɵ13, ɵ14 };
